airflow/models/xcom_arg.py

```diff
@@ -309,11 +310,19 @@ def zip(self, *others: XComArg, fillvalue: Any = NOTSET) -> ZipXComArg:
         return super().zip(*others, fillvalue=fillvalue)
 
     def get_task_map_length(self, run_id: str, *, session: Session) -> int | None:
+        from airflow.models.taskinstance import TaskInstance
         from airflow.models.taskmap import TaskMap
         from airflow.models.xcom import XCom
 
         task = self.operator
         if isinstance(task, MappedOperator):
+            state_query = session.query(TaskInstance.state).filter(
+                TaskInstance.dag_id == task.dag_id,
+                TaskInstance.run_id == run_id,
+                TaskInstance.task_id == task.task_id,
+            )
+            if any(state in State.unfinished for state, in state_query):
+                return None  # Not all of the expanded tis are done yet.
```

Comment:
Why not getting the count of not finished tasks or the first one if it exists instead of loading all the mapped tasks from the db and doing it with python?
something like:
```suggestion
            unfinished_ti = session.query(TaskInstance).filter(
                TaskInstance.dag_id == task.dag_id,
                TaskInstance.run_id == run_id,
                TaskInstance.task_id == task.task_id,
                TaskInstance.state.in_(State.unfinished),
            ).first()
            if unfinished_ti is not None:
                return None  # Not all of the expanded tis are done yet.
```

Reply:
I just tried this, and for some reason it does not work! It always returns 0 for me (no unfinished ti) even though there is clearly one ti with state None. I’m not sure what the problem is, perhaps something to do with NULL or string coersion?

Reply:
maybe we just need to convert `State.unfinished` to a list or a list of string :thinking:

Reply:
Tried that as well, still no luck.

It seems that NULL is actually special when used with IN, so we’ll need to do some special handling. I’ll try figure out how to write that clause. https://stackoverflow.com/questions/6362112/in-clause-with-null-or-is-null

Reply:
Ah I see, an alternative simple solution could be using `TaskInstance.state.not_in(State.finished)`

Reply:
Haven’t tried, but if `NULL` is the cultrip, `NOT IN` would not work either, because if `state` is NULL, `state NOT IN finished` would be `FALSE`, which is still incorrect.